### ISSUE TITLE
[Meta] Port existing GatsbyJS code to Next.js

### ISSUE THREAD

**Issue by humphd** [Seneca-CDOT/telescope#1316]
## Co-ordination of the GatsbyJS to Next.js Port

In #1314 we are beginning the transition from [GatsbyJS](https://www.gatsbyjs.com/) to [Next.js](https://nextjs.org/) for the front-end.  We will run parallel versions of the front-end while the transition occurs, with the current GatsbyJS version being the one we ship.

A new branch, [next](https://github.com/Seneca-CDOT/telescope/tree/next), has been created to facilitate this work in parallel to the `master` branch.  **All PRs for the Next.js port should happen against the `next` vs. `master` branch**.

We are going to be introducing [TypeScript](https://www.typescriptlang.org/) in the front-end.

This issue will remain open for the duration of the port, and we'll use it to organize what is done and what is still happening.

## How to Participate

We'd love to have you help with the work!  If you'd like to take on some part of the port, please let us know by commenting in this issue, and mentioning which of the tasks you'd like to take.

When you have chosen a task, this comment will get updated to indicate that you're working on it.  You should file an issue for the work, and we'll include that issue number in this comment.  When it's done, we'll indicate that in this comment.

When doing your work, please consult the file(s) in [the PR that removed the Gatsby code](https://github.com/Seneca-CDOT/telescope/pull/1314/files) as well as the [current front-end in `master`](https://github.com/Seneca-CDOT/telescope/tree/master/src/frontend), which may have evolved.

When you work on a task, you should try to decide if the old code is still needed.  Some things might only have been necessary when using GatsbyJS.  You should also make sure that we aren't missing something specific to Next.js.

Please see [the GatsbyJS to Next.js migration guide](https://nextjs.org/docs/migrating/from-gatsby)

## Source Code Note

In https://github.com/Seneca-CDOT/telescope/commit/6195af3fb3b9a277ca52e24a72c95e765be57791 we merged the `next` branch with `master`, giving us two frontends in the same tree:

- `src/frontend/gatsby` is our current Gatsby.js based frontend.  Please don't break it, as we are still running it on https://dev.telescope.cdot.systems (our staging box).
- `src/frontend/next` is our new Next.js based frontend.  You should add new code here.

See https://github.com/Seneca-CDOT/telescope/blob/master/src/frontend/README.md for details on how to use the two front-ends.

## Live Preview

The next.js version of the `master` branch is available at https://telescope-next.netlify.app/.  You can also use PR Previews to see what's changed in GitHub.

## Tasks

This is a high-level list of tasks that need to be addressed:

- [ ] Dynamic image from Unsplash Collection, see src/frontend/gatsby-node.js
- [ ] Integrate Material UI with Next.js, see https://github.com/mui-org/material-ui/tree/master/examples/nextjs-with-typescript 
- [ ] Expose `telescopeUrl` from `env` to frontend
- [ ] Figure out PR Previews using Vercel for Seneca-CDOT/next branch (#1447)
- [x] Port our front-end tests to use Next.js style  (**hold off until decision on separating Next into another repo**)
- [x] Port Components from Gatsby to Next.js:
    - [x] Deal with global config of pages/styles, see src/frontend/src/pages/PageBase.js, src/frontend/src/styles/globals.css, src/frontend/src/templates/template.js ([Pedro Fonseca](https://github.com/Seneca-CDOT/telescope/issues/1548))
    - [x] Theme src/frontend/src/theme.js ([Tony Vu](https://github.com/Seneca-CDOT/telescope/issues/1386))
    - [x] Main page, see src/frontend/src/pages/index.js ([Pedro Fonseca](https://github.com/Seneca-CDOT/telescope/issues/1358))
    - [x] AboutFooter/ ([Abdulbasid Guled](https://github.com/Seneca-CDOT/telescope/issues/1454))
    - [x] AdminButtons
    - [x] AuthorResult/ ([Joel Azwar](https://github.com/Seneca-CDOT/telescope/issues/1478)) (Unused/Outdated)
    - [x] Banner/  ([Abdulbasid Guled](https://github.com/Seneca-CDOT/telescope/issues/1454))
    - [x] BackToTopButton/ ([Joel Azwar](https://github.com/Seneca-CDOT/telescope/issues/1480))
    - [x] DynamicImage/
    - [x] GitHubContributorCard/ ([Joel Azwar](https://github.com/Seneca-CDOT/telescope/issues/1496))
    - [x] Footer/  ([Abdulbasid Guled](https://github.com/Seneca-CDOT/telescope/issues/1454))
    - [x] Header/   ([Jie Yang](https://github.com/Seneca-CDOT/telescope/issues/1446))
    - [x] Header/DesktopHeader   ([Jie Yang](https://github.com/Seneca-CDOT/telescope/issues/1446))
    - [x] Header/MobileHeader  ([Jie Yang](https://github.com/Seneca-CDOT/telescope/issues/1446))
    - [x] Login/ ([Joey Assal](https://github.com/Seneca-CDOT/telescope/issues/1456))
    - [x] LogoIcon/ ([Chris Pinkney](https://github.com/Seneca-CDOT/telescope/issues/1481))
    - [x] MyFeedsPage/ ([Tony Vu](https://github.com/Seneca-CDOT/telescope/issues/1497))
    - [x] Post/ ([Roger Yu](https://github.com/Seneca-CDOT/telescope/issues/1459))
    - [x] Posts/ ([Roger Yu](https://github.com/Seneca-CDOT/telescope/issues/1459))
    - [x] SEO.js/ ([Hyunji Lee](https://github.com/Seneca-CDOT/telescope/issues/1508))
    - [x] ScrollAction/ ([Junyong Liu](https://github.com/Seneca-CDOT/telescope/issues/1469))
    - [x] SearchBar/ (Pedro Fonseca)
    - [X] SearchHelp/([Hyunji Lee](https://github.com/Seneca-CDOT/telescope/issues/1507))
    - [x] SearchInput/(Pedro Fonseca)
    - [x] SearchPage/ ([Calvin Ho](https://github.com/Seneca-CDOT/telescope/issues/1438))
    - [x] SnackBar/ <— can we delete this now?
    - [x] Spinner/ ([Calvin Ho](https://github.com/Seneca-CDOT/telescope/issues/1479))
    - [x] User/
    - [x] hooks/
    - [x] About page, see src/frontend/src/markdown-pages/about.md ([Chris Pinkney](https://github.com/Seneca-CDOT/telescope/issues/1463))
    - [x] Error page, see src/frontend/src/pages/error.jsx ([Hyunji Lee](https://github.com/Seneca-CDOT/telescope/issues/1468))
    - [x] MyFeedsPage, see src/frontend/src/pages/myfeeds.js ([Tony Vu](https://github.com/Seneca-CDOT/telescope/issues/1491))
    - [x] SearchPage, see src/frontend/src/pages/search.js ([Tony Vu](https://github.com/Seneca-CDOT/telescope/issues/1484))

**Comment by c3ho** [comment]
Hi, I would like to work on SearchPage

**Comment by humphd** [comment]
@c3ho file an issue for it, edit the top comment beside `SearchPage/` to include the link to the issue number and your name.

**Comment by PedroFonsecaDEV** [comment]
Hi,
I would like to work on issues related to: Port Components from Gatsby to Next.js

**Comment by humphd** [comment]
@PedroFonsecaDEV pick one of the items in the list, file an issue, and edit comment at the top with a link to the issue and your name.

**Comment by jossiey** [comment]
I would like to work on Header, including Desktop header and Mobile header

**Comment by HyperTHD** [comment]
I shall attempt to port the banner and footer pages including the about footer page if everyone is okay with that

**Comment by humphd** [comment]
@HyperTHD perfect.  Can you file two issues, then edit the comment at the top of this issue to include your name and issue numbers?

**Comment by joey2031** [comment]
Hi, I can I do Login/?

**Comment by humphd** [comment]
I need reviews on #1444 so I can unleash you all on this work.

**Comment by rogercyyu** [comment]
Hi, can I do `Post/ ` & `Posts/`?

**Comment by humphd** [comment]
@rogercyyu sure, please follow the same advice I've given above about how to assign yourself to these.

**Comment by humphd** [comment]
## UPDATE

In https://github.com/Seneca-CDOT/telescope/commit/6195af3fb3b9a277ca52e24a72c95e765be57791 we merged the `next` branch with `master`, giving us two frontends in the same tree:

- `src/frontend/gatsby` is our current Gatsby.js based frontend.  Please don't break it, as we are still running it on https://dev.telescope.cdot.systems (our staging box).
- `src/frontend/next` is our new Next.js based frontend.  You should add new code here.

See https://github.com/Seneca-CDOT/telescope/blob/master/src/frontend/README.md for details on how to use the two front-ends.

You should update your `master` and re-run `npm install`, then remove the old Gatsby build directories we don't need anymore:

```
git checkout master
git pull upstream master
rm -fr src/frontend/.cache
rm -fr src/frontend/public
rm -fr src/frontend/node_modules
```

**Comment by hyunjiLeeTech** [comment]
Hi, I can I take Error page part for it?

**Comment by humphd** [comment]
@hyunjiLeeTech sure, follow advice above for signing up.  NOTE: the error page stuff is partly done in the backend in Express and partly in frontend.  @raygervais has worked on this, and can likely provide details when you get further along.

**Comment by MSTEWARDSON** [comment]
Can I take on the SearchBar conversion? If I'm reading this write I need to open an issue for it, then edit this page to mark I'm working on it correct?

**Comment by hyunjiLeeTech** [comment]
Hi, if I would like to take AdminButton, how can I check whether my work is properly working or not? I ask this question because when I see the Gatsby code for AdminButton, it seems that is checks whether the user is admin or not. Please tell me if it would be better to take another task.

**Comment by humphd** [comment]
Closing and we'll do final work in #1790
